synapse/rest/client/v2_alpha/report_event.py

```diff
@@ -58,17 +81,108 @@ async def on_POST(self, request, room_id, event_id):
                 Codes.BAD_JSON,
             )
 
+        target = body.get("org.matrix.msc2938.target", "homeserver-admins")
+
+        if target == "room-moderators":
+            # Report event to room moderators.
+            # This branch has further safety checks (e.g. can the user actually see the event?)
+            return await self._target_room_moderators(room_id=room_id, event_id=event_id, user_id=user_id, reason=body["reason"], score=body["score"])
+        elif target == "homeserver-admins":
+            return await self._target_homeserver_admin(room_id=room_id, event_id=event_id, user_id=user_id, reason=body["reason"], content=body)
+        else:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "Optional param 'target' must be one of ['homeserver-admins', 'room-moderators']",
+                Codes.BAD_JSON,
+            )
+
+    async def _target_homeserver_admin(self, room_id, event_id, user_id, reason, content):
+        # Store event report for later investigation by the homeserver admin.
         await self.store.add_event_report(
             room_id=room_id,
             event_id=event_id,
             user_id=user_id,
-            reason=body["reason"],
-            content=body,
+            reason=reason,
+            content=content,
             received_ts=self.clock.time_msec(),
         )
 
         return 200, {}
 
+    async def _target_room_moderators(self, room_id, event_id, user_id, reason, score):
```

Comment:
Please add docstrings and type hints.

This seems *very* different than how we handle the ones for homeserver admins. Does this behavior make sense?

In particular note that the server notices system is disabled by default in Synapse.

In particular is there need for any more state to be kept of these reports (e.g. the admin one gets stored into a table).

Reply:
> Please add docstrings and type hints.
> 
> This seems _very_ different than how we handle the ones for homeserver admins. Does this behavior make sense?

Well, this is being summarized/discussed [here](https://github.com/matrix-org/matrix-doc/pull/2938). It seems to be the best way for moderators to actually receive the info.

> In particular note that the server notices system is disabled by default in Synapse.

It is enabled on matrix.org and EMS, though, right?

> In particular is there need for any more state to be kept of these reports (e.g. the admin one gets stored into a table).

Not as far as I can tell. It's a simple message. We can't do that for the homeserver admin because the homeserver admin doesn't have an account, but in the future, we could imagine changing how it works for admins.